image-hd.c

```diff
@@ -282,8 +287,59 @@ static int hdimage_insert_gpt(struct image *image, struct list_head *partitions)
 		table[i].flags = part->bootable ? GPT_PE_FLAG_BOOTABLE : 0;
 		for (j = 0; j < strlen(part->name) && j < 36; j++)
 			table[i].name[j] = htole16(part->name[j]);
+
+		if (part->partition_type) {
+			struct mbr_partition_entry *entry;
+
+			entry = (struct mbr_partition_entry *)(part_table + i *
+				sizeof(struct mbr_partition_entry));
+
+			entry->boot = part->bootable ? 0x80 : 0x00;
+
+			entry->partition_type = part->partition_type;
+			entry->relative_sectors = part->offset/512;
+			entry->total_sectors = part->size/512;
+
+			hdimage_setup_chs(entry->relative_sectors, entry->first_chs);
+			hdimage_setup_chs(entry->relative_sectors +
+			entry->total_sectors - 1, entry->last_chs);
+
+			hybrid++;
+		}
 		i++;
 	}
+
+	if (hybrid) {
+		struct mbr_partition_entry *entry;
+
+		entry = (struct mbr_partition_entry *)(part_table + i *
+			sizeof(struct mbr_partition_entry));
+
+		entry->boot = 0x00;
+
+		entry->partition_type = 0xee;
+		entry->relative_sectors = 1;
+		entry->total_sectors = (512+sizeof(header)+sizeof(table))/512;
+
+		hdimage_setup_chs(entry->relative_sectors, entry->first_chs);
+		hdimage_setup_chs(entry->relative_sectors +
+		entry->total_sectors - 1, entry->last_chs);
+
+		image_info(image, "writing Hybrid MBR\n");
+
+		part_table += 4 * sizeof(struct mbr_partition_entry);
+		part_table[0] = 0x55;
+		part_table[1] = 0xaa;
+
+		ret = insert_data(image, mbr, imageoutfile(image), sizeof(mbr), 440);
+		if (ret) {
+			image_error(image, "failed to write Hybrid MBR\n");
+			return ret;
+		}
+	}
```

Comment:
This looks like a lot of code duplication. I think it would be better to reuse `hdimage_insert_mbr()` and modify that a bit.
For example, skip partitions if `(hybrid && !part->partition_type)` and things like this.

Reply:
I attempted that approach first actually but got a bit stuck, I don't really write a whole lot of c in general so I may need a bit of help with the adaptation.

Reply:
What was the problem. I'd just count the hybrid partition like you do now and and the end do:
```
if (hybrid)
		ret = hdimage_insert_mbr(image);
else
		ret = hdimage_insert_protective_mbr(image);
```
And skip some partitions as mentioned above. Or did you change anything else. I've not looked that closely at the new code.

Reply:
Ah, I think I had actually tried to modify `hdimage_insert_protective_mbr()` and not `hdimage_insert_mbr()`, I'll see if I can get that working. I'm not sure how well your example would work since the hybrid partitions are detected as part of the mbr partition loop. With how I have it currently working it's not known if we are in hybrid mode until we finish iterating through the partitions.

Reply:
No, you detect the hybrid partitions at the beginning of `hdimage_insert_gpt()`, so it can be used at the end to decide if a protective or real mbr should be used.

Reply:
I've reworked this to use `hdimage_insert_mbr()` for the hybrid partition table.